### ISSUE TITLE
🔋🤝🗒️   Support for EV-initiated Renegotiation & Fine Grain Charging Profiles

### ISSUE THREAD

**Issue by the-bay-kay** [EVerest/everest-demo#71]
## Context
This PR builds off of the work done in Issue #69.  These changes include:
- The addition of ISO 15118-2's "DepartureTime" (DT) and "EAmount"
- The Integration of an LQR Optimizer to optimize charging sessions

Currently, the LQR Optimizer generates a full schedule from timestamp 0 to DT, and takes a uniform sample of 24 - (SECC_Schedule) points to generate the charging profile.  This is not optimal for a variety of reasons -- as such, we are exploring options for updating schedules with a finer resolution. 

## High Level Goals
ISO 15118-2 defines a protocol for EV initiated Re-Negotiation of charging profiles in several places, such as [V2G2-521] and [V2G2-813].  I've created a diagram outlining a successful re-negotiation, attached below:

<details>
<summary> Renegotiation Sequence Diagram </summary>

<img width="612" alt="image" src="https://github.com/user-attachments/assets/692a7bb6-d967-4411-8dc3-05a48897f7fa">
</details>

Inspecting [iso15118_2_states.py](https://github.com/EVerest/ext-switchev-iso15118/blob/7f16c4b2c1307ce73798215f34b8fe06862bbae1/iso15118/evcc/states/iso15118_2_states.py#L1216C14-L1216C28), it appears we have support for this re-negotation, but do not include any new charging profiles.  As such, the goal of this issue is to integrate the Optimizer script into this function, such that we can re-negotation regularly (e.g., every 24 seconds, or another fine-grain amount) and more effectively optimize our charging session.

### Subsequent Goals
The following are an approximation of the work we will need to do in order to support these changes. 
- [x] Within the simulator, we will need to calculate our current timestamp relative to the start time, and ISO 15118-2's departure time.
  - I have yet to investigate where this change will occur, but @shankari mentioned she may be able to shed some light on this! 
- [x] We must adjust the optimizer scripts such that it returns a schedule in linear 24-bit chunks.
  - E.g.: Given a granularity of 1 second, and we've charged for 4 minutes (240 seconds), we'd return the powercurve points for timestamp 241-265
- [x] We will need to adjust the SASchedule integration such that the fine-grain Profile Schedule is compatible with any PMax values set
  - We will find the set of schedule entries that fall within (timestamp,  timestamp + 24*[granularity), and "weave in" the profile schedule between these values
- [x] Finally, we will a call to the optimizer script within [iso15118_2_states.py](https://github.com/EVerest/ext-switchev-iso15118/blob/7f16c4b2c1307ce73798215f34b8fe06862bbae1/iso15118/evcc/states/iso15118_2_states.py#L1216C14-L1216C28)'s `ChargingStatus` Class.  This profile will then be sent to the SECC via the existing PowerDeliveryRequest

 Additionally, there are some clean-ups we would like to perform:
- Currently, the Node-RED UI shows our charger always at "max power".  This was a hack to get around a known Node-RED bug, which has since been patched.  We would like to 
  - [x] Update Node-RED to the latest version, and
  - [ ] Revert the changes to the power meter.
- [x] Currently, the demo works in real time.  While this is useful in terms of accuracy, it makes it difficult to showcase our functions in real time (e.g., running a full demo of a 3-Hour departure time would be quite boring...). We hope to add a user input to adjust the simulation timescale, so that we can run demos at 5x / 10x / 100x speed.

**Comment by the-bay-kay** [comment]
> Even in (2), we need to know what the current time is, so that we can check to see if it has passed.
As far as I can tell, this is maintained in self.comm_session.ongoing_timer

I definitely agree that (2) will be the simplest method of implementation.  I'm not 100% that the comm session handler is exactly what we're looking for, however -- during the ChargingStatusRes, it appears the timer isn't active:

<img width="1421" alt="image" src="https://github.com/user-attachments/assets/0fcdf457-d58d-4f5b-8751-a037a1b69b00">

Is this timer then to check for EVCC <-> SECC communication timeouts?  I'll do a little digging to see what exactly it's for.  Regardless, like you said, this is a good example for how we could implement our own timer!

**Comment by the-bay-kay** [comment]
We've got a way to track and check time now.  We should be able to store the end of our active ProfileScheduleEntry and trigger a re-negotiation once we reach that timestamp.  This will require a rewrite of the schedule "weaving" algorithm, let's tackle that after we add the timestamp storage. 

<img width="371" alt="image" src="https://github.com/user-attachments/assets/9631a8d5-35a7-44fa-a82e-c4f46ee89c33">

It should be straight forward to add speed up functionality to our current clock.  I'm following the `ongoing_timer`'s model by (i) storing the timestamp of initialization, and (ii) calculating the time elapsed between that value and each `ChargingStatusRes`.  Our renegotiation will look something like...

```python
speed_factor = mqtt_fetch(value) # 
time_elapsed = (time() * speed_up_factor) - self.comm_session.charging_timer
if (time_elapsed > self.comm_session.end_of_schedule):
    # Renegotiate...
```

**Comment by shankari** [comment]
> time_elapsed = (time() * speed_up_factor) - self.comm_session.charging_timer

Nice! So we don't have to use `libfaketime`, we can just handle it directly

**Comment by the-bay-kay** [comment]
@shankari , the demo is committed!  Took a bit longer than expected to handle the end-of-schedule behavior, but you should be able to run the demo as you did the [previous](https://github.com/EVerest/everest-demo/issues/69#issuecomment-2369178032), substituting for [this](https://github.com/the-bay-kay/everest-demo/tree/demo-renegotiation) branch.  For your convenience...

```bash
git clone https://github.com/the-bay-kay/everest-demo.git; cd everest-demo; git checkout demo-renegotiation; bash demo-iso15118-2-ac-plus-ocpp.sh -r $(pwd) -b test-demo -3
```

**Comment by the-bay-kay** [comment]
Some good news: Looking at the latest version of the [ISO 15118-2 DC charging](https://github.com/EVerest/everest-demo/?tab=readme-ov-file#step-1-run-the-demo), it appears they have already added a method for viewing the ISO15118-2 logs -- we should be be able to use this as a template (or use it in its entirety) to create our own version of these (or expand them to OCPP 2.0.1!)

![image](https://github.com/user-attachments/assets/5d32f11a-eda3-4f04-864c-bb3b5cf430db)

**Comment by shankari** [comment]
Yes, that was my prior attempt to have AFS display the logs in node-red (aka "they" is not the ongoing EVerest community).
However, it displays the ISO messages as encoded EXI and does not display OCPP messages at all.

**Comment by the-bay-kay** [comment]
It seems that updating the Power Gauge may be more complicated than updating the functions in Node-RED.  Looking at the flows, we can see that we are receiving *nothing* from the `/powermeter/totalKWattHr/` MQTT publication... 

https://github.com/user-attachments/assets/8fafdd4a-f4f1-4819-8919-ad63a4b713b7

... compared to the latest demos, in which we *are* receiving these values.  So, let's look at the two modules where this message could be published:
- YetiDriver/index.js ✘
  - If this comment is to be believed, these publications are [depracated](https://github.com/EVerest/everest-core/blob/b11c690bc2158b713a16c5e81be73341a94cfd74/modules/simulation/JsYetiSimulator/index.js#L921)... So, let's ignore this and assume that
- EvseManager/EvseManager.cpp ✔
  - Given the above information, this is likely to be where our publication is coming from!
  
So, what's the difference between our demos and the working power gauge from main?  Well -- a lot, actually.  Looking at the [blame](https://github.com/EVerest/everest-core/blame/main/modules/EvseManager/ErrorHandling.cpp) history, there have been multiple changes to this module since we branched off...

I think there's two options for how we can proceed forward from here:
- Spend more time de-bugging our old MQTT messages -- we can look at the broker's logs, re-compile to add some statements into the publications, etc.
- We proceed ahead with [rolling forward](https://github.com/EVerest/everest-demo/issues/74) the demo to the latest stable release (ideally September, TBD): once we do so, the updates to `EvseManager.cpp` should clear this up

I am currently leaning pretty strongly to the second option.  Even if this doesn't fix everything (I'll almost certainly need to tweak the flows, for example), updating before debugging will at least ensure that we're using a contemporary version of this module.  That way, whatever work I do de-bugging will not be "useless" after roll-forward.  If we go ahead with the rolling forward, I will close this issue (as my current demo images _are_ working!) and move discussion to the thread linked above.
